### PR TITLE
Search suggestions

### PR DESCRIPTION
**Pull request by ghiitais**

### CODE REVIEW

File: RunningOrder/Shared/Components/Search/SearchBarSuggestions+Logic.swift [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+import Foundation
+import Combine
+import SwiftUI
+
+extension SearchBarSuggestions {
+    final class Logic: ObservableObject {
+        @Binding var input: String
+        private unowned var storyManager: StoryManager
+        private unowned var sprintManager: SprintManager
+        var searchManager: SearchManager
```
Comment (zarghol) [inline RunningOrder/Shared/Components/Search/SearchBarSuggestions+Logic.swift]: this one is not private or unowned ?

File: RunningOrder/Shared/Components/Search/SearchBarSuggestions+Logic.swift [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        var filteredStories: [Story] {
+            var wholeStories: [Story] = []
+            sprintManager.sprints.forEach { sprint in
+                let storiesPerId = storyManager.stories[sprint.id] ?? []
+                wholeStories.append(contentsOf: storiesPerId)
+            }
+
+            return wholeStories.filter { story -> Bool in
+                return story.name.lowercased().contains(input.lowercased()) || story.epic.lowercased().contains(input.lowercased()) || story.ticketReference.lowercased().contains(input.lowercased())
+            }
```
Comment (zarghol) [inline RunningOrder/Shared/Components/Search/SearchBarSuggestions+Logic.swift]:
```suggestion
            sprintManager.sprints
               .flatMap { sprint in return storyManager.stories[sprint.id] ?? [] }
               .filter { story -> Bool in
                   return story.name.lowercased().contains(input.lowercased()) || story.epic.lowercased().contains(input.lowercased()) || story.ticketReference.lowercased().contains(input.lowercased())
            }
```

could something like this be better ?

File: RunningOrder/Shared/Components/Search/SearchBarSuggestions+Logic.swift [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+                                          items: formattedStories.removingDuplicates()),
+                            SearchSection(name: SearchSection.SectionType.epic.rawValue.uppercased(),
+                                          items: filteredEpics.removingDuplicates())]
+
+            return sections
+        }
+    }
+}
+
+extension Array where Element: Hashable {
```
Comment (zarghol) [inline RunningOrder/Shared/Components/Search/SearchBarSuggestions+Logic.swift]: interesting extension :) could be added in another file, or file private if you want to keep it here.

File: RunningOrder/Shared/Components/Search/SearchBarSuggestions+Logic.swift [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    func removingDuplicates() -> [Element] {
+        var addedDict = [Element: Bool]()
+
+        return filter {
```
Comment (zarghol) [inline RunningOrder/Shared/Components/Search/SearchBarSuggestions+Logic.swift]: interesting approach. what do you think of using a Set instead ? I don't know if it is more optimized or not..

File: RunningOrder/Shared/Components/Search/SearchBarSuggestions.swift [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+extension SearchBarSuggestions {
+    fileprivate struct InternalView: View {
+        @ObservedObject var logic: Logic
+
+        var body: some View {
+            ScrollView {
+                VStack(alignment: .leading) {
+                    if !logic.filteredStories.isEmpty {
+                        ForEach(logic.filteredSearchSections, id: \.id) { section in
+                            Section(header: Text(section.name).font(.headline).bold()) {
```
Comment (zarghol) [inline RunningOrder/Shared/Components/Search/SearchBarSuggestions.swift]: What do you think of creating a new view 
```swift
extension SearchBarSuggestions {
   struct SectionHeader: View {
      //...
   }
}
```
Do you think it is worthing it, or not ?

File: RunningOrder/Shared/Components/Search/SearchBarSuggestions.swift [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+        InternalView(logic: Logic(input: $searchText, storyManager: storyManager, sprintManager: sprintManager, searchManager: searchManager))
+    }
+}
+
+extension SearchBarSuggestions {
+    fileprivate struct InternalView: View {
+        @ObservedObject var logic: Logic
+
+        var body: some View {
+            ScrollView {
```
Comment (zarghol) [inline RunningOrder/Shared/Components/Search/SearchBarSuggestions.swift:26]: Why using a ScrollView instead of List ?

File: RunningOrder/Shared/Components/Search/SearchBarView.swift [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+                            inputText = ""
+                            disableTextField = true
+                        })
+                    TextField(searchManager.isItemSelected ? "" : "Search", text: $inputText).disabled(disableTextField)
+
+                } else {
+                    TextField(searchManager.isItemSelected ? "" : "Search", text: $inputText)
+                }
+            }
+
```
Comment (zarghol) [inline RunningOrder/Shared/Components/Search/SearchBarView.swift:30]: remove space :)

File: RunningOrder/Shared/Components/Search/SearchBarView.swift [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+    var body: some View {
+        HStack {
+            ZStack(alignment: .leading) {
+                if let selected = searchManager.selectedSearchItem?.name {
+                    Tag("\(selected)", color: Color(identifier: .gray).opacity(0.25), foregroundTextColor: Color.black)
+                        .padding(.trailing, 18)
+                        .onAppear(perform: {
+                            inputText = ""
+                            disableTextField = true
```
Comment (zarghol) [inline RunningOrder/Shared/Components/Search/SearchBarView.swift:25]: so, we can only have one tag for now ?

File: RunningOrder/Shared/Components/Search/SearchBarView.swift [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+                        })
+                    TextField(searchManager.isItemSelected ? "" : "Search", text: $inputText).disabled(disableTextField)
```
Comment (zarghol) [inline RunningOrder/Shared/Components/Search/SearchBarView.swift]: this could be outside the if let { } and not having else :)

File: RunningOrder/Shared/Components/Search/SearchBarView.swift [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+
+                } else {
+                    TextField(searchManager.isItemSelected ? "" : "Search", text: $inputText)
+                }
+            }
+
+            .overlay(
+                HStack {
+                    Spacer()
+                    if !inputText.isEmpty || disableTextField {
```
Comment (zarghol) [inline RunningOrder/Shared/Components/Search/SearchBarView.swift:34]: so I the overlay, in case of text empty and textfield enabled, you won't have the x button, ok, but why do you have the HStack ? Maybe consider if { ... } else { Empty() } or something like it

File: RunningOrder/Shared/Components/Search/SearchBarView.swift [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+                        Button(action: {
+                            self.inputText = ""
```
Comment (zarghol) [inline RunningOrder/Shared/Components/Search/SearchBarView.swift:36]: I see lots of actions that can be in a Logic of this view. Have you considered to create one Logic ?

File: RunningOrder/Shared/MainView.swift [hunk 12]
```diff
@@ -66,7 +66,7 @@ extension MainView {
                                 Spacer()
                             }
 
-                            ToolbarItem(placement: ToolbarItemPlacement.cancellationAction) {
+                            ToolbarItemGroup(placement: ToolbarItemPlacement.cancellationAction) {
```
Comment (zarghol) [inline RunningOrder/Shared/MainView.swift]: you just changed the item to group but did't add yours ?

File: RunningOrder/Sprint/Models/Sprint.swift [hunk 13]
```diff
@@ -23,3 +23,40 @@ extension Sprint {
 
 extension Sprint: Equatable { }
 extension Sprint: Hashable { }
+
+struct SearchSection: Identifiable {
+    var id: UUID { return UUID() }
```
Comment (zarghol) [inline RunningOrder/Sprint/Models/Sprint.swift]:
```suggestion
    var id: UUID = UUID()
```

why a computed ? don't you want id to be fixed ?
If this is only for conforming to Identifiable, the goal is to have SearchSection unique : can two sections have the same name in the view ? I don't think so consider having
`var id: String { name }` in an extension for conformance

File: RunningOrder/Sprint/Models/Sprint.swift [hunk 14]
```diff
@@ -23,3 +23,40 @@ extension Sprint {
...
+        }
+    }
+}
+
+struct SearchItem: Identifiable, Hashable {
+    static func == (lhs: SearchItem, rhs: SearchItem) -> Bool {
+        lhs.name == rhs.name
+    }
+
+    var id: UUID { return UUID() }
```
Comment (zarghol) [inline RunningOrder/Sprint/Models/Sprint.swift]: same : add conformance in an extension, not with a computed if you create a new not fixed value in it

you made SeachItem Equatable by name, why not using name in the id ?

File: RunningOrder/Sprint/Models/Sprint.swift [hunk 15]
```diff
@@ -23,3 +23,40 @@ extension Sprint {
...
 extension Sprint: Hashable { }
+
+struct SearchSection: Identifiable {
+    var id: UUID { return UUID() }
+    var name: String
+    var items: [SearchItem]
+}
+
+extension SearchSection {
+    enum SectionType: String {
```
Comment (zarghol) [inline RunningOrder/Sprint/Models/Sprint.swift]: good !! 👍  why not using directly the enum as SearchSection ? :)

File: RunningOrder/Sprint/Views/SprintList.swift [hunk 16]
```diff
@@ -14,13 +14,18 @@ extension Sprint: Identifiable {}
...
+        @State private var searchText: String = ""
+        var searchSections: [SearchSection] = []
         let space: Space
 
         var body: some View {
             List {
+                SearchBarView(inputText: $searchText)
+
                 Section(header: Text("Active Sprints")) {
+                    // show view
```
Comment (zarghol) [inline RunningOrder/Sprint/Views/SprintList.swift]: is this comment useful ? 🤔

File: RunningOrder/Sprint/Views/SprintList.swift [hunk 17]
```diff
@@ -14,13 +14,18 @@ extension Sprint: Identifiable {}
...
+        @EnvironmentObject var storyManager: StoryManager
         @ObservedObject var logic: Logic
-
+        @State private var searchText: String = ""
+        var searchSections: [SearchSection] = []
         let space: Space
 
         var body: some View {
             List {
+                SearchBarView(inputText: $searchText)
```
Comment (zarghol) [inline RunningOrder/Sprint/Views/SprintList.swift]: why do you have the search bar here instead of Tool bar ?

File: RunningOrder/Story/Views/StoryList+Logic.swift [hunk 18]
```diff
@@ -42,5 +50,40 @@ extension StoryList {
 
             return Color.Identifier.epicColors[epicIndex]
         }
+
+        var allStories: [Story] {
+            var wholeStories: [Story] = []
+            sprintManager.sprints.forEach { sprint in
```
Comment (zarghol) [inline RunningOrder/Story/Views/StoryList+Logic.swift]: we shouldn't access to all the existing stories in this logic object as we only display these for the current sprint (and furthermore these for the current Space !)

File: RunningOrder/Story/Views/StoryList+Logic.swift [hunk 19]
```diff
@@ -42,5 +50,40 @@ extension StoryList {
...
+                let storiesPerId = storyManager.stories[sprint.id] ?? []
+                wholeStories.append(contentsOf: storiesPerId)
+            }
+
+            return wholeStories
+        }
+
+        /// Returns the stories of a specific sprintId
+        /// - Parameter sprintId: The id of the sprint
+        func stories(for sprintId: Sprint.ID, filter: String? = nil) -> [Story] {
```
Comment (zarghol) [inline RunningOrder/Story/Views/StoryList+Logic.swift]: same as above, It is the manager job to give the logic the stories for the current sprint

File: RunningOrder/Story/Views/StoryList.swift [hunk 20]
```diff
@@ -16,57 +16,59 @@ extension StoryList {
...
-            .navigationTitle("Sprint \(sprint.number) - \(sprint.name)")
-            .frame(minWidth: 100, idealWidth: 300)
-            .toolbar {
-                ToolbarItems.sidebarItem
+            .navigationTitle(logic.isSearchFound ? "Searching" : "Sprint \(sprint.number) - \(sprint.name)")
+                .frame(minWidth: 100, idealWidth: 300)
+                .toolbar {
+                    ToolbarItems.sidebarItem
+
+                    ToolbarItemGroup(placement: ToolbarItemPlacement.cancellationAction) {
```
Comment (zarghol) [inline RunningOrder/Story/Views/StoryList.swift]: why change from item to group ?

File: RunningOrder/Story/Views/StoryList.swift [hunk 21]
```diff
@@ -16,57 +16,59 @@ extension StoryList {
...
+                        }
                     }
                 }
-            }
-            .navigationTitle("Sprint \(sprint.number) - \(sprint.name)")
-            .frame(minWidth: 100, idealWidth: 300)
-            .toolbar {
-                ToolbarItems.sidebarItem
+            .navigationTitle(logic.isSearchFound ? "Searching" : "Sprint \(sprint.number) - \(sprint.name)")
+                .frame(minWidth: 100, idealWidth: 300)
```
Comment (zarghol) [inline RunningOrder/Story/Views/StoryList.swift]: please reindent

File: RunningOrder/Story/Managers/StoryManager.swift [hunk 22]
```diff
@@ -38,11 +37,6 @@ final class StoryManager: ObservableObject {
         }).store(in: &cancellables)
     }
 
-    /// Returns the stories of a specific sprintId
-    /// - Parameter sprintId: The id of the sprint
-    func stories(for sprintId: Sprint.ID) -> [Story] {
```
Comment (zarghol) [inline RunningOrder/Story/Managers/StoryManager.swift:43]: :/

**Review by zarghol — CHANGES_REQUESTED** [review]
Nice work, but some points need attention

File: RunningOrder/Shared/MainView.swift (hunk 12 above)
Comment (ghiitais) [reply RunningOrder/Shared/MainView.swift]: old change, i'll revert it

File: RunningOrder/Sprint/Views/SprintList.swift (hunk 16 above)
Comment (ghiitais) [reply RunningOrder/Sprint/Views/SprintList.swift]: no removed

File: RunningOrder/Shared/Components/Search/SearchBarSuggestions+Logic.swift (hunk 3 above)
Comment (ghiitais) [reply RunningOrder/Shared/Components/Search/SearchBarSuggestions+Logic.swift]: done, moved to Sequence + Extensions

File: RunningOrder/Shared/Components/Search/SearchBarSuggestions.swift (hunk 6 above)
Comment (ghiitais) [reply RunningOrder/Shared/Components/Search/SearchBarSuggestions.swift:26]: I had a weird behavior with it ..

File: RunningOrder/Shared/Components/Search/SearchBarSuggestions+Logic.swift (hunk 1 above)
Comment (ghiitais) [reply RunningOrder/Shared/Components/Search/SearchBarSuggestions+Logic.swift]: it should be unowned! but I can't have it private, i need to access the selected item

File: RunningOrder/Story/Views/StoryList.swift (hunk 20 above)
Comment (ghiitais) [reply RunningOrder/Story/Views/StoryList.swift]: reverted

File: RunningOrder/Shared/Components/Search/SearchBarView.swift (hunk 10 above)
Comment (ghiitais) [reply RunningOrder/Shared/Components/Search/SearchBarView.swift:34]: HStack is to host the Spacer and the button

File: RunningOrder/Shared/Components/Search/SearchBarSuggestions+Logic.swift (hunk 4 above)
Comment (ghiitais) [reply RunningOrder/Shared/Components/Search/SearchBarSuggestions+Logic.swift]: very good idea!

File: RunningOrder/Sprint/Models/Sprint.swift (hunk 15 above)
Comment (ghiitais) [reply RunningOrder/Sprint/Models/Sprint.swift]: becasue I need it to be a struct to contain the property `items`

File: RunningOrder/Shared/Components/Search/SearchBarSuggestions+Logic.swift [hunk 23]
```diff
@@ -0,0 +1,47 @@
...
+
+        init(input: Binding<String>, storyManager: StoryManager, searchManager: SearchManager) {
+            self._input = input
+            self.storyManager = storyManager
+            self.searchManager = searchManager
+        }
+
+        var filteredStories: [Story] {
+            let stories = storyManager.allStories.filter {$0.name.lowercased().contains(input.lowercased()) || $0.epic.lowercased().contains(input.lowercased()) || $0.ticketReference.lowercased().contains(input.lowercased())
+            }
```
Comment (zarghol) [inline RunningOrder/Shared/Components/Search/SearchBarSuggestions+Logic.swift:27]: could you re indent please ? :)

File: RunningOrder/Shared/Components/Search/SearchBarView.swift (hunk 11 above)
Comment (zarghol) [reply RunningOrder/Shared/Components/Search/SearchBarView.swift:36]: still no logic ?

File: RunningOrder/Shared/Components/Search/SearchBarView.swift (hunk 10 above)
Comment (zarghol) [reply RunningOrder/Shared/Components/Search/SearchBarView.swift:34]: ah yes overlay don't have a @ViewBuilder, it takes directly to View. okay

File: RunningOrder/Shared/MainView.swift [hunk 24]
```diff
@@ -12,6 +12,7 @@ extension MainView {
     private struct InternalView: View {
         @EnvironmentObject var spaceManager: SpaceManager
         @ObservedObject var logic: Logic
+        @State private var searchText: String = ""
```
Comment (zarghol) [inline RunningOrder/Shared/MainView.swift]: this seems weird as this is not related to this view at all.. why not embed it in the SearchManager ?

File: RunningOrder/Sprint/Views/SprintList.swift [hunk 25]
```diff
@@ -14,12 +14,14 @@ extension Sprint: Identifiable {}
 extension SprintList {
     fileprivate struct InternalView: View {
         @EnvironmentObject var sprintManager: SprintManager
+        @EnvironmentObject var storyManager: StoryManager
         @ObservedObject var logic: Logic
-
+//        @State private var searchText: String = ""
```
Comment (zarghol) [inline RunningOrder/Sprint/Views/SprintList.swift]: remove comment ? :)

File: RunningOrder/Story/Managers/StoryManager.swift [hunk 26]
```diff
@@ -138,6 +131,35 @@ final class StoryManager: ObservableObject {
             }
         }
     }
+
+    var allStories: [Story] {
```
Comment (zarghol) [inline RunningOrder/Story/Managers/StoryManager.swift:135]: you need it at only one place, maybe remove this var and replace the caller with `stories.values.flatMap { $0 }`directly :)

File: RunningOrder/Story/Views/StoryDetail.swift [hunk 27]
```diff
@@ -13,6 +13,7 @@ struct StoryDetail: View {
 
     @EnvironmentObject var storyInformationManager: StoryInformationManager
     @EnvironmentObject var storyManager: StoryManager
+    @State private var searchText: String = ""
```
Comment (zarghol) [inline RunningOrder/Story/Views/StoryDetail.swift]: As said for the mainView, maybe add it to the SearchManager

File: RunningOrder/Story/Views/StoryList.swift [hunk 28]
```diff
@@ -43,12 +42,13 @@ extension StoryList {
                     }
                 }
             }
-            .navigationTitle("Sprint \(sprint.number) - \(sprint.name)")
+            .navigationTitle(logic.isSearchFound ? "Searching" : "Sprint \(sprint.number) - \(sprint.name)")
```
Comment (zarghol) [inline RunningOrder/Story/Views/StoryList.swift]: why `logic.isSearchFound` instead of `searchManager.isItemSelected` ?

File: RunningOrder/Story/Views/StoryList.swift [hunk 29]
```diff
@@ -43,12 +42,13 @@ extension StoryList {
...
             .frame(minWidth: 100, idealWidth: 300)
             .toolbar {
                 ToolbarItems.sidebarItem
 
                 ToolbarItem(placement: ToolbarItemPlacement.cancellationAction) {
+
```
Comment (zarghol) [inline RunningOrder/Story/Views/StoryList.swift]: remove space ? :)

File: RunningOrder/Story/Views/StoryList.swift [hunk 30]
```diff
@@ -64,9 +64,11 @@ extension StoryList {
 struct StoryList: View {
     let sprint: Sprint
     @EnvironmentObject var storyManager: StoryManager
+    @EnvironmentObject var searchManager: SearchManager
+    @EnvironmentObject var sprintManager: SprintManager
```
Comment (zarghol) [inline RunningOrder/Story/Views/StoryList.swift]: this seems to be not used anymore. you can remove it :)

**Review by zarghol — APPROVED** [review]

File: RunningOrder/Story/Managers/StoryManager.swift (hunk 26 above)
Comment (ghiitais) [reply RunningOrder/Story/Managers/StoryManager.swift:135]: it is also called in SearchBarSuggestions+Logic

File: RunningOrder/Story/Views/StoryList.swift (hunk 28 above)
Comment (ghiitais) [reply RunningOrder/Story/Views/StoryList.swift]: yeah ^^' i'll rename

**Review by Neckow — APPROVED** [review]

**Comment by zarghol** [comment]
Merged. As we need it tomorrow. As you already are on some more fixes, I won't create a new ticket for them. Feel free to create one if you want :)
